File: CSharp/ILinkai.Weixin.Sdk/WeixinApiClient.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace ILinkai.Weixin.Sdk;

/// <summary>
/// ILinkai微信API客户端配置选项
/// </summary>
public class WeixinApiClientOptions
{
    /// <summary>
    /// API基础URL
    /// </summary>
    public string BaseUrl { get; set; } = "https://ilinkai.weixin.qq.com";

    /// <summary>
    /// CDN基础URL
    /// </summary>
    public string CdnBaseUrl { get; set; } = "https://novac2c.cdn.weixin.qq.com/c2c";

    /// <summary>
    /// 认证令牌
    /// </summary>
    public string? Token { get; set; }

    /// <summary>
    /// 请求超时时间（毫秒），默认15000
    /// </summary>
    public int TimeoutMs { get; set; } = 15000;

    /// <summary>
    /// 长轮询超时时间（毫秒），默认35000
    /// </summary>
    public int LongPollTimeoutMs { get; set; } = 35000;

    /// <summary>
    /// 配置请求超时时间（毫秒），默认10000
    /// </summary>
    public int ConfigTimeoutMs { get; set; } = 10000;

    /// <summary>
    /// 路由标签
    /// </summary>
    public string? RouteTag { get; set; }
}

/// <summary>
/// ILinkai微信API客户端
/// 提供与ilinkai.weixin.qq.com API交互的核心功能
/// </summary>
public class WeixinApiClient
{
    private readonly WeixinApiClientOptions _options;
    private readonly HttpClient _httpClient;
    private readonly ILogger? _logger;
    private readonly string _channelVersion;
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
    };

    /// <summary>
    /// 默认API基础URL
    /// </summary>
    public const string DefaultBaseUrl = "https://ilinkai.weixin.qq.com";

    /// <summary>
    /// 默认CDN基础URL
    /// </summary>
    public const string DefaultCdnBaseUrl = "https://novac2c.cdn.weixin.qq.com/c2c";

    /// <summary>
    /// 会话过期错误码
    /// </summary>
    public const int SessionExpiredErrorCode = -14;

    /// <summary>
    /// 创建WeixinApiClient实例
    /// </summary>
    /// <param name="options">配置选项</param>
    /// <param name="logger">日志记录器</param>
    public WeixinApiClient(WeixinApiClientOptions? options = null, ILogger? logger = null)
    {
        _options = options ?? new WeixinApiClientOptions();
        _logger = logger;
        _channelVersion = GetChannelVersion();
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromMilliseconds(Math.Max(_options.TimeoutMs, _options.LongPollTimeoutMs) + 5000)
        };
    }

    /// <summary>
    /// 获取渠道版本号
    /// </summary>
    private static string GetChannelVersion()
    {
        try
        {
            var version = typeof(WeixinApiClient).Assembly.GetName().Version;
            return version?.ToString() ?? "1.0.0";
        }
        catch
        {
            return "unknown";
        }
    }

    /// <summary>
    /// 构建基础信息
    /// </summary>
    private Models.BaseInfo BuildBaseInfo()
    {
        return new Models.BaseInfo
        {
            ChannelVersion = _channelVersion
        };
    }

    /// <summary>
    /// 生成随机微信UIN头部值
    /// </summary>
    private static string GenerateRandomWechatUin()
    {
        var randomBytes = new byte[4];
        RandomNumberGenerator.Fill(randomBytes);
        var uint32Value = BitConverter.ToUInt32(randomBytes, 0);
        var decimalString = uint32Value.ToString();
        return Convert.ToBase64String(Encoding.UTF8.GetBytes(decimalString));
    }

    /// <summary>
    /// 构建请求头
    /// </summary>
    private Dictionary<string, string> BuildHeaders(string body)
    {
        var headers = new Dictionary<string, string>
        {
            ["Content-Type"] = "application/json",
            ["AuthorizationType"] = "ilink_bot_token",
            ["Content-Length"] = Encoding.UTF8.GetByteCount(body).ToString(),
            ["X-WECHAT-UIN"] = GenerateRandomWechatUin()
        };

        if (!string.IsNullOrWhiteSpace(_options.Token))
        {
            headers["Authorization"] = $"Bearer {_options.Token.Trim()}";
        }

        if (!string.IsNullOrWhiteSpace(_options.RouteTag))
        {
            headers["SKRouteTag"] = _options.RouteTag;
        }

        return headers;
    }

    /// <summary>
    /// 确保URL以斜杠结尾
    /// </summary>
    private static string EnsureTrailingSlash(string url)
    {
        return url.EndsWith("/") ? url : $"{url}/";
    }

    /// <summary>
    /// 发送API请求
    /// </summary>
    private async Task<string> ApiFetchAsync(string endpoint, object body, int timeoutMs, string label)
    {
        var baseUrl = EnsureTrailingSlash(_options.BaseUrl);
        var url = $"{baseUrl}{endpoint}";
        var jsonBody = JsonSerializer.Serialize(body, JsonOptions);
        var headers = BuildHeaders(jsonBody);

        _logger?.LogDebug("POST {Url} body={Body}", RedactUrl(url), RedactBody(jsonBody));

        using var request = new HttpRequestMessage(HttpMethod.Post, url);
        request.Content = new StringContent(jsonBody, Encoding.UTF8, "application/json");

        foreach (var header in headers)
        {
            if (!request.Headers.TryAddWithoutValidation(header.Key, header.Value))
            {
                request.Content?.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }
        }

        using var cts = new CancellationTokenSource(timeoutMs);
        var response = await _httpClient.SendAsync(request, cts.Token);
        var rawText = await response.Content.ReadAsStringAsync(cts.Token);

        _logger?.LogDebug("{Label} status={Status} raw={Raw}", label, (int)response.StatusCode, RedactBody(rawText));

        if (!response.IsSuccessStatusCode)
        {
            throw new WeixinApiException($"{label} {(int)response.StatusCode}: {rawText}", (int)response.StatusCode);
        }

        return rawText;
    }

    /// <summary>
    /// 获取更新（长轮询）
    /// </summary>
    /// <param name="getUpdatesBuf">更新缓冲区</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>更新响应</returns>
    public async Task<Models.GetUpdatesResponse> GetUpdatesAsync(string? getUpdatesBuf = null, CancellationToken cancellationToken = default)
    {
        var request = new Models.GetUpdatesRequest
        {
            GetUpdatesBuf = getUpdatesBuf ?? string.Empty,
            BaseInfo = BuildBaseInfo()
        };

        try
        {
            var rawText = await ApiFetchAsync("ilink/bot/getupdates", request, _options.LongPollTimeoutMs, "getUpdates");
            return JsonSerializer.Deserialize<Models.GetUpdatesResponse>(rawText, JsonOptions)
                ?? new Models.GetUpdatesResponse();
        }
        catch (OperationCanceledException)
        {
            _logger?.LogDebug("getUpdates: client-side timeout after {Timeout}ms, returning empty response", _options.LongPollTimeoutMs);
            return new Models.GetUpdatesResponse
            {
                Ret = 0,
                Messages = new List<Models.WeixinMessage>(),
                GetUpdatesBuf = getUpdatesBuf
            };
        }
    }

    /// <summary>
    /// 获取上传URL
    /// </summary>
    /// <param name="request">请求参数</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>上传URL响应</returns>
    public async Task<Models.GetUploadUrlResponse> GetUploadUrlAsync(Models.GetUploadUrlRequest request, CancellationToken cancellationToken = default)
    {
        request.BaseInfo = BuildBaseInfo();
        var rawText = await ApiFetchAsync("ilink/bot/getuploadurl", request, _options.TimeoutMs, "getUploadUrl");
        return JsonSerializer.Deserialize<Models.GetUploadUrlResponse>(rawText, JsonOptions)
            ?? new Models.GetUploadUrlResponse();
    }

    /// <summary>
    /// 发送消息
    /// </summary>
    /// <param name="request">发送消息请求</param>
    /// <param name="cancellationToken">取消令牌</param>
    public async Task SendMessageAsync(Models.SendMessageRequest request, CancellationToken cancellationToken = default)
    {
        if (request.Message != null)
        {
            request.Message = new Models.WeixinMessage
            {
                Seq = request.Message.Seq,
                MessageId = request.Message.MessageId,
                FromUserId = request.Message.FromUserId,
                ToUserId = request.Message.ToUserId,
                ClientId = request.Message.ClientId,
                CreateTimeMs = request.Message.CreateTimeMs,
                UpdateTimeMs = request.Message.UpdateTimeMs,
                DeleteTimeMs = request.Message.DeleteTimeMs,
                SessionId = request.Message.SessionId,
                GroupId = request.Message.GroupId,
                MessageType = request.Message.MessageType,
                MessageState = request.Message.MessageState,
                ItemList = request.Message.ItemList,
                ContextToken = request.Message.ContextToken
            };
        }
        await ApiFetchAsync("ilink/bot/sendmessage", request, _options.TimeoutMs, "sendMessage");
    }

    /// <summary>
    /// 获取配置
    /// </summary>
    /// <param name="ilinkUserId">ILink用户ID</param>
    /// <param name="contextToken">上下文令牌</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>配置响应</returns>
    public async Task<Models.GetConfigResponse> GetConfigAsync(string ilinkUserId, string? contextToken = null, CancellationToken cancellationToken = default)
    {
        var request = new Models.GetConfigRequest
        {
            ILinkUserId = ilinkUserId,
            ContextToken = contextToken,
            BaseInfo = BuildBaseInfo()
        };
        var rawText = await ApiFetchAsync("ilink/bot/getconfig", request, _options.ConfigTimeoutMs, "getConfig");
        return JsonSerializer.Deserialize<Models.GetConfigResponse>(rawText, JsonOptions)
            ?? new Models.GetConfigResponse();
    }

    /// <summary>
    /// 发送输入状态
    /// </summary>
    /// <param name="request">输入状态请求</param>
    /// <param name="cancellationToken">取消令牌</param>
    public async Task SendTypingAsync(Models.SendTypingRequest request, CancellationToken cancellationToken = default)
    {
        request.BaseInfo = BuildBaseInfo();
        await ApiFetchAsync("ilink/bot/sendtyping", request, _options.ConfigTimeoutMs, "sendTyping");
    }

    /// <summary>
    /// 脱敏URL
    /// </summary>
    private static string RedactUrl(string url)
    {
        try
        {
            var uri = new Uri(url);
            var query = uri.Query;
            if (string.IsNullOrEmpty(query)) return url;
            return url.Split('?')[0] + "?[REDACTED]";
        }
        catch
        {
            return url;
        }
    }

    /// <summary>
    /// 脱敏请求体
    /// </summary>
    private static string RedactBody(string body)
    {
        if (string.IsNullOrEmpty(body) || body.Length < 200) return body;
        return body.Substring(0, 200) + "...[TRUNCATED]";
    }
}

/// <summary>
/// 微信API异常
/// </summary>
public class WeixinApiException : Exception
{
    /// <summary>
    /// HTTP状态码
    /// </summary>
    public int StatusCode { get; }

    /// <summary>
    /// 创建WeixinApiException实例
    /// </summary>
    public WeixinApiException(string message, int statusCode) : base(message)
    {
        StatusCode = statusCode;
    }
}

File: Python/ilinkai_weixin/weixin_api_client.py
"""ILinkai微信API客户端"""
import json
import logging
import os
import struct
import base64
from dataclasses import dataclass, field
from typing import Optional
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError

from .models import (
    BaseInfo, GetUpdatesRequest, GetUpdatesResponse, GetUploadUrlRequest,
    GetUploadUrlResponse, SendMessageRequest, GetConfigRequest, GetConfigResponse,
    SendTypingRequest, WeixinMessage, to_dict, parse_weixin_message,
)

logger = logging.getLogger(__name__)


class WeixinApiException(Exception):
    def __init__(self, message: str, status_code: int = 0):
        super().__init__(message)
        self.status_code = status_code


@dataclass
class WeixinApiClientOptions:
    base_url: str = "https://ilinkai.weixin.qq.com"
    cdn_base_url: str = "https://novac2c.cdn.weixin.qq.com/c2c"
    token: Optional[str] = None
    timeout_ms: int = 15000
    long_poll_timeout_ms: int = 35000
    config_timeout_ms: int = 10000
    route_tag: Optional[str] = None

DEFAULT_BASE_URL = "https://ilinkai.weixin.qq.com"
DEFAULT_CDN_BASE_URL = "https://novac2c.cdn.weixin.qq.com/c2c"
SESSION_EXPIRED_ERROR_CODE = -14


class WeixinApiClient:
    """ILinkai微信API客户端"""

    def __init__(self, options: Optional[WeixinApiClientOptions] = None):
        self.options = options or WeixinApiClientOptions()
        self._channel_version = self._get_channel_version()

    @staticmethod
    def _get_channel_version() -> str:
        return "1.0.0"

    def _build_base_info(self) -> BaseInfo:
        return BaseInfo(channel_version=self._channel_version)

    @staticmethod
    def _generate_random_wechat_uin() -> str:
        random_bytes = os.urandom(4)
        uint32_value = struct.unpack('<I', random_bytes)[0]
        decimal_string = str(uint32_value)
        return base64.b64encode(decimal_string.encode('utf-8')).decode('ascii')

    @staticmethod
    def _ensure_trailing_slash(url: str) -> str:
        return url if url.endswith('/') else url + '/'

    def _build_headers(self, body: bytes) -> dict:
        headers = {
            'Content-Type': 'application/json',
            'AuthorizationType': 'ilink_bot_token',
            'Content-Length': str(len(body)),
            'X-WECHAT-UIN': self._generate_random_wechat_uin(),
        }
        if self.options.token and self.options.token.strip():
            headers['Authorization'] = f'Bearer {self.options.token.strip()}'
        if self.options.route_tag and self.options.route_tag.strip():
            headers['SKRouteTag'] = self.options.route_tag
        return headers

    def _api_fetch(self, endpoint: str, body: object, timeout_ms: int, label: str) -> str:
        base_url = self._ensure_trailing_slash(self.options.base_url)
        url = f'{base_url}{endpoint}'
        json_body = json.dumps(to_dict(body), ensure_ascii=False)
        body_bytes = json_body.encode('utf-8')
        headers = self._build_headers(body_bytes)

        logger.debug("POST %s body=%s", _redact_url(url), _redact_body(json_body))

        req = Request(url, data=body_bytes, headers=headers, method='POST')
        timeout_sec = timeout_ms / 1000.0

        try:
            with urlopen(req, timeout=timeout_sec) as resp:
                raw_text = resp.read().decode('utf-8')
                logger.debug("%s status=%s raw=%s", label, resp.status, _redact_body(raw_text))
                return raw_text
        except HTTPError as e:
            raw_text = e.read().decode('utf-8') if e.fp else ''
            raise WeixinApiException(f'{label} {e.code}: {raw_text}', e.code)

    def get_updates(self, get_updates_buf: Optional[str] = None) -> GetUpdatesResponse:
        """获取更新（长轮询）"""
        request = GetUpdatesRequest(
            get_updates_buf=get_updates_buf or '',
            base_info=self._build_base_info(),
        )
        try:
            raw_text = self._api_fetch('ilink/bot/getupdates', request,
                                        self.options.long_poll_timeout_ms, 'getUpdates')
            data = json.loads(raw_text)
            resp = GetUpdatesResponse(
                ret=data.get('ret'), errcode=data.get('errcode'), errmsg=data.get('errmsg'),
                get_updates_buf=data.get('get_updates_buf'),
                longpolling_timeout_ms=data.get('longpolling_timeout_ms'),
            )
            if data.get('msgs'):
                resp.messages = [parse_weixin_message(m) for m in data['msgs']]
            return resp
        except (URLError, TimeoutError):
            logger.debug("getUpdates: client-side timeout after %dms", self.options.long_poll_timeout_ms)
            return GetUpdatesResponse(ret=0, messages=[], get_updates_buf=get_updates_buf)

    def get_upload_url(self, request: GetUploadUrlRequest) -> GetUploadUrlResponse:
        """获取上传URL"""
        request.base_info = self._build_base_info()
        raw_text = self._api_fetch('ilink/bot/getuploadurl', request,
                                    self.options.timeout_ms, 'getUploadUrl')
        data = json.loads(raw_text)
        return GetUploadUrlResponse(
            upload_param=data.get('upload_param'),
            thumb_upload_param=data.get('thumb_upload_param'),
        )

    def send_message(self, request: SendMessageRequest) -> None:
        """发送消息"""
        self._api_fetch('ilink/bot/sendmessage', request, self.options.timeout_ms, 'sendMessage')

    def get_config(self, ilink_user_id: str, context_token: Optional[str] = None) -> GetConfigResponse:
        """获取配置"""
        request = GetConfigRequest(
            ilink_user_id=ilink_user_id,
            context_token=context_token,
            base_info=self._build_base_info(),
        )
        raw_text = self._api_fetch('ilink/bot/getconfig', request,
                                    self.options.config_timeout_ms, 'getConfig')
        data = json.loads(raw_text)
        return GetConfigResponse(
            ret=data.get('ret'), errmsg=data.get('errmsg'),
            typing_ticket=data.get('typing_ticket'),
        )

    def send_typing(self, request: SendTypingRequest) -> None:
        """发送输入状态"""
        request.base_info = self._build_base_info()
        self._api_fetch('ilink/bot/sendtyping', request, self.options.config_timeout_ms, 'sendTyping')


def _redact_url(url: str) -> str:
    idx = url.find('?')
    return url[:idx] + '?[REDACTED]' if idx >= 0 else url

def _redact_body(body: str) -> str:
    if not body or len(body) < 200:
        return body
    return body[:200] + '...[TRUNCATED]'
